### PR TITLE
Experiment caching TermLike hashes

### PR DESCRIPTION
**Pull request by treeowl**
Cache the hash of the `TermLikeF` in the `TermLike`, and extract it in the `Hashable` instance instead of recursing into the `TermLikeF`.

The precomputed hash in produced using the derived `Hashable` instance of `TermLikeF`, which will call `hash` on the respective sub-structure (like `And`, `Or`, ...), wrapped into one of the `TermLikeF` constructors.  For all such sub-structures, the derived `Hashable` instance will then recurse into the contents (containing a `TermLike` again for the case of `TermLike`), thereby using the cached hashes directly instead of performing a deep recursion.

Partially addresses #3113

Fixes #nnnn

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

**Comment by treeowl** [comment]
There's a problem: `mapVariables` and `traverseVariables` generally need `Hashable variable2` to produce their results. However, when `variable1 ~ Concrete`, there are no variables in the input, which (I think?) means there won't be any in the output either. So it should be possible to write similar functions that don't put any constraints on the result variable type. Unfortunately, I don't understand enough about what's going on in `traverseVariables` to know how to adapt it properly.

It would be good to get this figured out, because `mapVariables` "infects" a bunch of other functions and instances with `Hashable` constraints that probably shouldn't be necessary.

**Comment by treeowl** [comment]
Hmmm... All the Nix jobs are failing. Any guesses?

File: kore/src/Kore/Internal/TermLike/TermLike.hs [hunk 1]
```diff
@@ -647,34 +648,31 @@ instance HasFreeVariables (TermAttributes variable) variable where
...
-    { getTermLike ::
-        CofreeF
-            (TermLikeF variable)
-            (TermAttributes variable)
-            (TermLike variable)
+data TermLike variable = TermLike__
+    { _tlAttributes :: TermAttributes variable -- Should this be strict?
+    -- The hash is stored lazily to avoid computing it unnecessarily. Do we generally
+    -- need the hash? If so, it should probably be strict.
+    , -- | A hash of _tlTermLikeF
```
Comment (jberthold) [inline kore/src/Kore/Internal/TermLike/TermLike.hs:657]: A quick heads-up about this: 
I would definitely prefer if the code was lazy by default but we currently have `Language Strict, StrictData` on by default and turn it of per module. (There is an experimentation ticket to switch to lazy and try to plug holes, to see the performance difference, but always too many things to do).

Comment (treeowl) [reply kore/src/Kore/Internal/TermLike/TermLike.hs:657]: Oof, that's very confusing!

**Comment by jberthold** [comment]
> Hmmm... All the Nix jobs are failing. Any guesses?

It is probably a GHC-9 specific error. Only nix jobs are testing GHC-9 in our CI setup (main build is with GHC-8.10.7 but we have the GHC-9 setup for profiling and to migrate to it at some point).
You could maybe test it by changing the resolver to what nix GHC-9 uses, see
https://github.com/runtimeverification/haskell-backend/blob/master/nix/kore-ghc923.nix.d/default.nix#L41

**Comment by goodlyrottenapple** [comment]
```
src/Kore/Internal/TermLike/TermLike.hs:695:10: error:
    • No instance for (Eq variable)
        arising from the superclasses of an instance declaration
      Possible fix:
        add (Eq variable) to the context of the instance declaration
    • In the instance declaration for ‘Hashable (TermLike variable)’
    |
695 | instance Hashable (TermLike variable) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

you can run `nix build .#kore-exec-ghc9 ` locally to check for any more issues with ghc9

**Comment by treeowl** [comment]
@goodlyrottenapple, I see. That's quite a mess. Has the `Hashable` change been dealt with elsewhere in this project? Is there an accepted approach?

**Comment by treeowl** [comment]
It seems GHC in Haddock mode has some sort of trouble with the CPP I need to support both versions of `hashable` we're currently using. Is there some local fix for that? Or can we bump the older `hashable` version to a suitable minimum?

**Comment by treeowl** [comment]
I have no idea how to interpret the remaining test failures.

**Comment by treeowl** [comment]
`vacuousVariables` really should be a no-op operationally, shouldn't it? `unsafeCoerce` would do? I imagine there are some reasons (formal verification?) not to do that, but it seems a bit sad to walk a whole term to change nothing at all.

**Comment by jberthold** [comment]
Profiling run with file from #3288 [kevm-bug-storagevar03-with-conditions.tar.gz](https://github.com/runtimeverification/haskell-backend/files/9670919/kevm-bug-storagevar03-with-conditions.tar.gz) (test)

**Comment by jberthold** [comment]
Profiling with file from #3286 : [first file, kevm-bug-storagevar03-with-conditions.tar.gz](https://github.com/runtimeverification/haskell-backend/files/9670919/kevm-bug-storagevar03-with-conditions.tar.gz)

**Comment by goodlyrottenapple** [comment]
> @goodlyrottenapple, I see. That's quite a mess. Has the `Hashable` change been dealt with elsewhere in this project? Is there an accepted approach?

Don't think its been an issue so far. We added GHC 9 a few months ago and I remember there were only a few changes necessary at the time to make the code compile for both, so as such, there is no apporach in place. We have an issue open to switch to GHC9, but need to go through our dependencies again and see if support for GHC9 has been up-streamed for the packages which weren't compatible at the time.

**Comment by goodlyrottenapple** [comment]
> It seems GHC in Haddock mode has some sort of trouble with the CPP I need to support both versions of `hashable` we're currently using. Is there some local fix for that? Or can we bump the older `hashable` version to a suitable minimum?

Think we can just bump `hashable` if it makes things easier.

**Comment by treeowl** [comment]
Closing in favor of #3338; this branch was off a fork, which apparently causes issues.

**Pull request closed without merging**